### PR TITLE
Stream RO-Crate Zip

### PR DESCRIPTION
**Pull request by dnlbauer**
This is a Draft implementation showing how #205 could be solved.

Implementation is rather straightforward for file entities where I could simply add a method to stream the file contents.
For datasets entities, it was a bit challenging to find a good solution. Creating a zip stream requires that all the file handling need to happen at one place where the zip stream object is created (`rocrate.stream_zip` in this case). However, dataset entities currently handle the file writing operation internally and don't expose their containing files in an easy to use way. I solved this with a generator that not only yields the data of the contained files, but also yields file paths. However, the solution still looks a bit "hacky" to me. I didn't want to do more refactoring though.

### CODE REVIEW

**Comment by dnlbauer** [comment]
How it works:


```python
#test_zipstream.py
from rocrate.rocrate import ROCrate

crate = ROCrate(gen_preview=True)

# add a file
testfile = crate.add_file("file.txt", properties={
    "name": "my test file",
}, record_size=True)

# add a folder
crate.add_dataset(
            source="folder",
            dest_path="internal_folder",
    )

# add a remote dataset
crate.add_dataset(
    source="https://raw.githubusercontent.com/ResearchObject/ro-crate-py/refs/heads/master/",
    dest_path="remote/folder/",
    validate_url=False,
    fetch_remote=True,
    properties={
        "hasPart": [{"@id": "README.md"}]
    }
)

# write zip stream to file.
# Instead, it could also be used as response to an url request
# to stream the content with very low memory footprint and no disk usage
with open("out.zip", "wb") as out:
    for chunk in crate.stream_zip():
        out.write(chunk)
```

```bash
$> python test_zipstream.py
$> unzip -l out.zip
Archive:  out.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
       19  1980-01-01 00:00   file.txt
        5  1980-01-01 00:00   internal_folder/test.txt
    17395  1980-01-01 00:00   remote/folder/README.md
     6027  1980-01-01 00:00   ro-crate-preview.html
     1371  1980-01-01 00:00   ro-crate-metadata.json
---------                     -------
    24817                     5 files
```

**Comment by dnlbauer** [comment]
Seems like I broke some test cases, but I think we can first discuss if this is an approach worth to continue pursuing before I start to clean things up.

**Comment by simleo** [comment]
I tried this:

```python
from rocrate.rocrate import ROCrate

in_crate_dir = "test/test-data/ro-crate-galaxy-sortchangecase"
out_crate_zip = "/tmp/ro-crate-galaxy-sortchangecase.zip"
crate = ROCrate(in_crate_dir)
with open(out_crate_zip, "wb") as f:
    for chunk in crate.stream_zip():
        f.write(chunk)
```

And the output crate is missing `test/test1/input.bed` and `test/test1/output_exp.bed`. They are missing even if I do `crate.write_zip(out_crate_zip)` instead of the loop with the new method. This is due to the fact that the rewritten `write_zip` does not call `write`, which in turn calls a method to copy files that are unlisted in the metadata. Looking at the changes, I see that several existing methods have been rewritten, hence the failure of existing tests. Could you manage to implement the new functionality without touching existing methods, or at least with only minimal changes to them? This way, the chance of breaking existing stuff would be much smaller.

**Comment by dnlbauer** [comment]
Yes there are some implementation because I tried to replace the write code. It's probably easier to, as you suggested, take a step back and implement it without changing the write code. This will lead to a lot of code duplication though.

**Comment by simleo** [comment]
> It's probably easier to, as you suggested, take a step back and implement it without changing the write code. This will lead to a lot of code duplication though.

Let's avoid code duplication then. Moving forward, please follow these steps:

1. Change the code so that all *current* tests pass, i.e., *without changing the test code*
2. Add new tests to test the new functionalities, making sure they pass as well
3. Check that this PR does not cause significant performance degradation

**Comment by dnlbauer** [comment]
I spent some time today to figure out why there are so many test failing. There were quite some edge-cases I missed which required some refactoring to adress.

### Updates:
- Writing an ro-crate as folder when the files are already present elsewhere now uses `shutil`, just like before this PR. This ensures there are no noticeable performance degradations for writing RO-Crates from existing files.
- Streaming is now only used when calling `stream`, or when writing files to disk which need to be streamed anyways (i.e. the input is an instance of `IOBase` or the data has to be fetched from remote)

## Things to do / to discuss:
- [ ] **Decide whether  `write_zip` should use streaming under the hood**: While this is optional, it comes with several advantages:
  - writing a zip to disk would not duplicate raw data anymore, which benefits systems with limited storage or systems where `/tmp` is mounted as RAM.
  - We can reuse the already existing unit tests for zip to test the streaming implementation. Otherwise,they will have to be implemented for streaming zips in addtion.
- [ ] **"extra data" handling does not work for streaming** -> There is no equivalent of `_copy_unlisted` for streaming yet.
- [ ] **Unit tests**

File: rocrate/model/dataset.py [hunk 1]
```diff
@@ -43,37 +44,84 @@ def _empty(self):
...
+                errno.ENOENT, os.strerror(errno.ENOENT), str(self.source)
+            )
+        if not self.crate.source:
+            for root, _, files in os.walk(self.source):
+                root = Path(root)
+                for name in files:
+                    source = root / name
+                    dest = source.relative_to(self.source.parent)
+                    with open(source, 'rb') as f:
+                        yield str(dest), f.read()
```
Comment (simleo) [inline rocrate/model/dataset.py:105]: `f.read()` reads the whole file in memory. Doesn't this defeat the whole purpose of being able to stream the crate in small chunks even if it contains huge files?

**Comment by simleo** [comment]
> I tried this:
> 
> ```python
> from rocrate.rocrate import ROCrate
> 
> in_crate_dir = "test/test-data/ro-crate-galaxy-sortchangecase"
> out_crate_zip = "/tmp/ro-crate-galaxy-sortchangecase.zip"
> crate = ROCrate(in_crate_dir)
> with open(out_crate_zip, "wb") as f:
>     for chunk in crate.stream_zip():
>         f.write(chunk)
> ```

I ran the above snippet again. There is something wrong with the zip that gets created:

```console
$ unzip -d ro-crate-galaxy-sortchangecase{,.zip}
Archive:  ro-crate-galaxy-sortchangecase.zip
warning [ro-crate-galaxy-sortchangecase.zip]:  245282 extra bytes at beginning or within zipfile
  (attempting to process anyway)
  inflating: ro-crate-galaxy-sortchangecase/sort-and-change-case.ga  
  inflating: ro-crate-galaxy-sortchangecase/LICENSE  
  inflating: ro-crate-galaxy-sortchangecase/README.md  
  inflating: ro-crate-galaxy-sortchangecase/test/test1/sort-and-change-case-test.yml  
  inflating: ro-crate-galaxy-sortchangecase/ro-crate-metadata.json  
```

File: rocrate/model/dataset.py (hunk 1 above)
Comment (dnlbauer) [reply rocrate/model/dataset.py:105]: good catch! this should probably also yield small chunks instead.

Comment (dnlbauer) [reply rocrate/model/dataset.py]: resolved with [c034c66](https://github.com/ResearchObject/ro-crate-py/pull/212/commits/c034c666281fd3237f4f87a9906bb41c8afce1ab)

**Comment by dnlbauer** [comment]
> I ran the above snippet again. There is something wrong with the zip that gets created:

My bad on not properly testing between two commits ... Seems like ZipFile somehow treats the BytesIO differently than a subclass from RawIOBase or sth like that. With a plain in memory buffer, it works.

File: rocrate/model/dataset.py [hunk 2]
```diff
@@ -43,37 +44,85 @@ def _empty(self):
...
+                errno.ENOENT, os.strerror(errno.ENOENT), str(self.source)
+            )
+        if not self.crate.source:
+            for root, _, files in os.walk(self.source):
+                root = Path(root)
+                for name in files:
+                    source = root / name
+                    dest = source.relative_to(Path(self.source).parent)
+                    with open(source, 'rb') as f:
+                        for chunk in f:
```
Comment (simleo) [inline rocrate/model/dataset.py:105]: This iterates over the file's "lines", even if it's opened in binary mode, i.e. with chunks delimited by `\n`. It should yield chunks of the same size (except for the last one) instead, independent from `\n` characters and reasonably large (I think you used 8192 elsewhere).

**Comment by simleo** [comment]
The chunks yielded by `stream_zip` are almost all empty and highly irregular in size:

```pycon
>>> from rocrate.rocrate import ROCrate
>>> in_crate_dir = "test/test-data/ro-crate-galaxy-sortchangecase"
>>> crate = ROCrate(in_crate_dir)
>>> chunks = [_ for _ in crate.stream_zip()]
>>> len(chunks)
310
>>> from collections import Counter
>>> Counter(len(_) for _ in chunks).most_common()
[(0, 304), (53, 1), (916, 1), (3607, 1), (348, 1), (168, 1), (1108, 1)]
>>> 
```

I was expecting a more regular stream, and not to see empty chunks.

**Comment by dnlbauer** [comment]
Good point. the reason for the non-uniform chunk sizes is not only because of the different input stream sizes, but also because the chunk size of chunks going into the zip is different then the output chunk sizes (compression, zip file format overhead, ..). So even if the input streams would yield nicely sized chunks, the sizes will be different from the output.

It shouldn't be too complicated to implement to wait for the buffer for a given chunk size to be filled before yielding, though. It might actually be sensible to make the chunk size a parameter that gets passed through to underlying read operations.

**Comment by dnlbauer** [comment]
- [bf1b990](https://github.com/ResearchObject/ro-crate-py/pull/212/commits/bf1b990f9acc6014a0702db37014ad8c50fa35da) added unlisted files from the crate to the stream.
- [5f0a709](https://github.com/ResearchObject/ro-crate-py/pull/212/commits/5f0a709a62e2f59fc9901661bf546b14f44211a1) switched `write_zip` to use the stream method. This makes all the unit tests that work on the zip to also test the streaming functionality.

File: rocrate/rocrate.py [hunk 3]
```diff
@@ -469,15 +470,55 @@ def write(self, base_path):
...
+                for root, dirs, files in walk(str(self.source), exclude=self.exclude):
+                    root = Path(root)
+                    for name in dirs:
+                        source = root / name
+                        dest = source.relative_to(self.source)
+                        dest.mkdir(parents=True, exist_ok=True)
+                    for name in files:
+                        source = root / name
+                        rel = source.relative_to(self.source)
+                        if not self.dereference(str(rel)) and not out_path.samefile(source):
```
Comment (simleo) [inline rocrate/rocrate.py:510]: This crashes if `out_path` is `None` (or anything that does not have a `samefile` attribute, actually). Moreover, an `out_path` argument does not make sense in `stream_zip`: there is no output path, you're just yielding chunks.

Comment (dnlbauer) [reply rocrate/rocrate.py]: **If** `write_zip` should use the streaming implementation, we need a way to pass the out_path to the streaming implementation since `write_zip` is currently intended to support the edge-case where the zip file is written into the ro-crate folder. Otherwise, this functionality (tested with the `test_no_zip_in_zip` test) will break.

What could be done is:
- Leave with `write_zip` with the implementation that writes the create and zips it using shutils
- Document that the `out_path` parameter is for internal use. (this is what I did)
- Hide the parameter by having an external wrapper around the stream method without the parameter.

I now switched to an external wrapper. Also added a unit test to test the streaming functionality when not called from `write_zip`, to make sure a crash like the one you mentioned are tested.

Comment (simleo) [reply rocrate/rocrate.py]: `stream_zip` is still not yielding unlisted files. See 6b33a90.

Comment (simleo) [reply rocrate/rocrate.py]: Note that only `test_stream` fails. `test_write_zip_copy_unlisted` passes.

**Comment by simleo** [comment]
I've made some adjustments after testing with the crate at https://zenodo.org/records/8095888, see [2178b42](https://github.com/ResearchObject/ro-crate-py/pull/212/commits/2178b420eff8aff00f23a9ab15ca89d7a91fce29).

@dnlbauer before merging I need you to update author and copyright info. Author info is in `CITATION.cff`, `setup.py` and `rocrate/__init__.py`. Copyright info is in `README.md`, `rocrate/__init__.py` and Python file headers.

**Comment by dnlbauer** [comment]
@simleo thanks! the fix for large files was important.
I updated the author list (assuming it's alphabetical by last-name from what was already there), and also added Senckenberg to the copyright statements.
